Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp

```diff
@@ -0,0 +1,149 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include <nlohmann/json.hpp>
+#include <stdexcept>
+#include <string>
+
+#include "Acts/Geometry/GeometryHierarchyMap.hpp"
+
+namespace Acts {
+
+/// Convert a geometry hierarchy map to/from Json.
+///
+/// @tparam value_t value type stored in the geometry hierarchy map
+///
+/// The value type is expected to be directly convertible to/from a Json object.
+/// It has to be either a fundamental type or appropriate `to_json` and
+/// `from_json` functions must be available.
+///
+/// A user-defined identifier is stored in the encoded Json object that is used
+/// to identify which value type is stored in the file. The identifier is
+/// checked for consistency when decoding the Json object.
+template <typename value_t>
+class GeometryHierarchyMapJsonConverter {
+ public:
+  using Value = value_t;
+  using Container = GeometryHierarchyMap<value_t>;
+
+  /// Construct the converter.
+  ///
+  /// @param valueIdentifier user-defined identifier for the stored value
+  GeometryHierarchyMapJsonConverter(std::string valueIdentifier)
+      : m_valueIdentifier(std::move(valueIdentifier)) {
+    if (m_valueIdentifier.empty()) {
+      throw std::invalid_argument("Value identifier must be non-empty");
+    }
+  }
+
+  /// Encode the geometry hierarchy map into a Json object.
+  ///
+  /// @param container Geometry hierarchy map that should be encoded
+  /// @return Encoded Json object
+  nlohmann::json toJson(const Container& container) const;
+
+  /// Decode a Json object into a geometry hierarchy map.
+  ///
+  /// @param encoded Json object that should be decoded
+  /// @return Decoded geometry hierarchy map
+  /// @throw std::invalid_argument in case of format errors
+  Container fromJson(const nlohmann::json& encoded) const;
+
+ private:
+  static constexpr const char* const kHeaderKey = "acts-geometry-hierarchy-map";
+  static constexpr const char* const kEntriesKey = "entries";
+  static constexpr int kFormatVersion = 0;
+
```

Comment:
Maybe there should be a way to set the Format Version ?

Reply:
The version can not be set by the user on purpose. The version is used to identify the file version by the parser. Imagine that we decide to tweak the format in the future, e.g. by changing the geometry id encoding. In that case, the updated converter will write a different version. Old converter can then identify that they can not parse the data and newer converters can use different implementations to handle different versions. This is all transparent to the user and letting him modify will most likely break this.

Reply:
Ok I understand now, format refer to the "Geometry hierarchy" format and you would update it directly in the converter if that were to change. I my mind it was related to the way you wrote the the value (for exemple if you changes the to_json you would change the format version). But this makes perfect sense. Would you mind adding a comment to explain that ? 
Something like : "Converter version, to be changed manual in case of Geometry Hierarchy change" ?

Reply:
Sure. Comment added.